win32/devdoc/socket_transport_win32.md

```diff
@@ -0,0 +1,296 @@
+
+If `WSARecv` return 0, `socket_transport_receive` shall do the following:
+
+- If `bytes_recv` is not `NULL`, `socket_transport_receive` shall copy the number of bytes into `bytes_recv`.
+
+- `socket_transport_receive` shall return `SOCKET_RECEIVE_OK`.
+
+If `WSARecv` returns an non-zero value, `socket_transport_receive` shall do the following:
+
+- If `WSAGetLastError` returns `WSA_IO_PENDING`, and `bytes_recv` is not `NULL`, `socket_transport_receive` shall set `bytes_recv` to 0.
+
+  - `socket_transport_receive` shall return `SOCKET_RECEIVE_WOULD_BLOCK`.
+
+- If `WSAGetLastError` does not returns `WSA_IO_PENDING` `socket_transport_receive` shall return `SOCKET_RECEIVE_ERROR`.
+
+`socket_transport_receive` shall call `sm_exec_end`.
+
+### socket_transport_listen
+
+```c
+MOCKABLE_FUNCTION(, int, socket_transport_listen, SOCKET_TRANSPORT_HANDLE, socket_transport, uint16_t, port);
+```
+
+`socket_transport_listen` is to listen for incoming connections.
+
+If `socket_transport` is `NULL`, `socket_transport_listen` shall fail and return a non-zero value.
+
+If `port` is `0`, `socket_transport_listen` shall fail and return a non-zero value.
+
+If the transport type is not `SOCKET_SERVER`, `socket_transport_listen` shall fail and return a non-zero value.
+
+`socket_transport_listen` shall call `sm_open_begin` to begin the open.
+
+If `sm_open_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_listen` shall fail and return a non-zero value.
+
+`socket_transport_listen` shall call `socket` with the params `AF_INET`, `SOCK_STREAM` and `IPPROTO_TCP`.
+
+`socket_transport_listen` shall bind to the socket by calling `bind`.
+
+`socket_transport_listen` shall start listening to incoming connection by calling `listen`.
+
+If successful `socket_transport_listen` shall call `sm_open_end` with `true`.
+
+If any failure is encountered, `socket_transport_listen` shall call `sm_open_end` with `false`, fail and return a non-zero value.
+
+### socket_transport_accept
+
+```c
+MOCKABLE_FUNCTION(, SOCKET_TRANSPORT_HANDLE, socket_transport_accept, SOCKET_TRANSPORT_HANDLE, socket_transport);
+```
+
+`socket_transport_accept` accepts the incoming connections.
+
+If `socket_transport` is `NULL`, `socket_transport_accept` shall fail and return a non-zero value.
+
+If the transport type is not `SOCKET_SERVER`, `socket_transport_accept` shall fail and return a non-zero value.
+
+`socket_transport_accept` shall call `sm_exec_begin`.
+
+If `sm_exec_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_accept` shall fail and return `SOCKET_SEND_ERROR`.
+
+`socket_transport_accept` shall call `select` determine if the socket is ready to be read passing a timeout of 10 milliseconds.
+
+`socket_transport_accept` shall call `accept` to accept the incoming socket connection.
+
+`socket_transport_accept` shall allocate a `SOCKET_TRANSPORT` for the incoming connection and call `sm_create` and `sm_open` on the connection.
```

Comment:
What is the transport type of the returned socket? I assume it's actually `SOCKET_CLIENT`, but should specify here (or if you take my other comment, this just returns a client type).

Is `SOCKET_CLIENT` the right name though? Isn't it really "SOCKET_LISTENER" and "SOCKET_DATA" or something like that? Not really a client/server difference, just a difference in whether you are sending/receiving or you are listening/accepting on the socket.

Reply:
Very true about the name.  I'll specify it in the spec.